Refactored PI-specific invoices

QuanMPhm opened the pull request:
Closes #69. The PI-specific is now implemented in the `PIInvoice` class. This class takes in the entire billable invoice and will export the PI invoices to local storage and S3. Because of its unique export file paths, it overrides the `export()`

Note that to follow the file path has changed slightly, from:
`f"{self.name}/{pi_instituition}_{pi}_{self.invoice_month}.csv"`
to...
`f"{self.name}/{pi_instituition}_{pi} {self.invoice_month}.csv"`

Note that replacement of the "_" to " ". I do this to be more consistent with the filename pattern for the other invoices.
Aside from that, I've also removed `upload_to_s3()` and its corresponding unit test, since it is no longer needed (after this PR, and certainly after all the refactor PRs are merged). @knikolla @naved001 Should I add a test for each invoice to see if they are exporting to desired S3 paths?

knikolla commented on process_report/invoices/pi_specific_invoice.py:
```diff
@@ -0,0 +1,38 @@
+        for pi in pi_list:
+            if pandas.isna(pi):
+                continue
+            pi_projects = self.data[self.data[invoice.PI_FIELD] == pi]
```
This seems like a great candidate to be split into a separate method that takes a pi as an argument and exports that PI only.

joachimweyl replied:
@QuanMPhm how goes the splitting of this section?

QuanMPhm replied:
It was resolved last week.

knikolla commented on process_report/invoices/pi_specific_invoice.py:
```diff
@@ -0,0 +1,38 @@
+        ):  # self.name is name of folder storing invoices
+            os.mkdir(self.name)
+
+        pi_list = self.data[invoice.PI_FIELD].unique()
```
This seems it should belong to `_prepare`, assigning a class attribute `self.pi_list`.

knikolla reviewed (changes requested):
A general note, try to split large sections of common code to separate clearly named functions or methods.

joachimweyl commented:
@knikolla please rereview.

QuanMPhm commented:
@knikolla @naved001 Since all the invoices now implement their own S3 export functions, this meant that the `upload_to_s3` function in `process_report.py` was [removed](https://github.com/CCI-MOC/process_csv_report/pull/73/files#diff-162d25340846c53be6b9c357f242c269b4564a175d9f873c6a37a63239dd5419L486-L497), along with its [test case](https://github.com/CCI-MOC/process_csv_report/pull/73/files#diff-2ab8a00ce6228ed90b8929a4e45622b1586da07acbc85aa2d71533747f594ca9L790-L820) to check that files were uploaded to their intended s3 filepaths. I wanted to ask if this means I should implement a new test case that checks the export function of every invoice to do the same test.

knikolla commented:
> @knikolla @naved001 Since all the invoices now implement their own S3 export functions, this meant that the `upload_to_s3` function in `process_report.py` was [removed](https://github.com/CCI-MOC/process_csv_report/pull/73/files#diff-162d25340846c53be6b9c357f242c269b4564a175d9f873c6a37a63239dd5419L486-L497), along with its [test case](https://github.com/CCI-MOC/process_csv_report/pull/73/files#diff-2ab8a00ce6228ed90b8929a4e45622b1586da07acbc85aa2d71533747f594ca9L790-L820) to check that files were uploaded to their intended s3 filepaths. I wanted to ask if this means I should implement a new test case that checks the export function of every invoice to do the same test.

You can write a test for the exporting functionality of the base class `Invoice`, but it's unnecessary for invoices that don't modify that in any way.

QuanMPhm commented:
@knikolla @naved001 Since the `export_s3` function does not need to be concerned about file extensions anymore, I have simplified the current test case for s3 exporting. I also updated the commit message to reflect this change.

naved001 commented on process_report/process_report.py:
```diff
@@ -262,17 +262,13 @@ def main():
-        for pi_invoice in os.listdir(args.output_folder):
-            invoice_list.append(os.path.join(args.output_folder, pi_invoice))
+    pi_inv = pi_specific_invoice.PIInvoice(
+        name=args.output_folder, invoice_month=invoice_month, data=billable_inv.data
```
so, for PI specific invoice you didn't pass in a copy of the billable_inv dataframe which is okay because we only export things for PIInvoice and no data is processed.

QuanMPhm replied:
Yes, that's correct
